## pandas 3 ejercicios

In [1]:
import numpy as np
import pandas as pd

### En la siguiente celda se guarda la información del excel

In [3]:
df1 = pd.read_excel("Data/Airbnb_NYC.xlsx", sheet_name = "NYC1")
df2 = pd.read_excel("Data/Airbnb_NYC.xlsx", sheet_name = "NYC2")

### Ejercicio 0:
- Une los dos DataFrames anteriores por la columna **`id`**. Guarda en nuevo DataFrame en **`df`**

In [8]:
df = pd.merge(left = df1, right = df2, on = "id", how = "inner")

### Ejercicio 1:
- Si quiero hospedarme en Manhattan una sola noche y para elegir el Airbnb me guio por el que tenga el mayor número de reviews, ¿donde me debería de hospedar?

In [14]:
df[df['municipio'] == "Manhattan"].sort_values(by = "number_of_reviews", ascending=False).head(1)

id                        name  host_id host_name  municipio  \
2031  903972  Great Bedroom in Manhattan  4734398        Jj  Manhattan   

      barrio  latitude  longitude     room_type  price  minimum_nights  \
2031  Harlem  40.82085  -73.94025  Private room   49.0             1.0   

      number_of_reviews last_review  reviews_per_month  \
2031              607.0  2019-06-21               7.75   

      calculated_host_listings_count  availability_365  
2031                             3.0             293.0

### Ejercicio 2:
- Busco un Airbnb en el municipio donde el precio promedio de alquiler sea el más bajo. Busco las hbitaciones donde la cantidad de review por mes sea el mayor. ¿Cuáles son las 3 habitaciones que mejor se adaptan a estos criterios?

In [24]:
municipio_mas_barato = df.groupby("municipio")['price'].mean().sort_values().index[0]

In [26]:
df[df['municipio'] == municipio_mas_barato].sort_values(by = "reviews_per_month", ascending=False).head(3)

id                                             name    host_id  \
23371  18914730                     IDEALLY LOCATED COZY COTTAGE    1776707   
41945  32595490  Peaceful 2 Bedroom apartment in a private house  245003299   
42252  32801214       TW #6 Private Suite 2nd Fl.  1 to 4 Guests  211136294   

            host_name municipio         barrio  latitude  longitude  \
23371            Luis     Bronx    Throgs Neck  40.83051  -73.82511   
41945         Yoselyn     Bronx   Clason Point  40.80924  -73.85107   
42252  Sharon & Erika     Bronx  Schuylerville  40.83560  -73.83369   

             room_type  price  minimum_nights  number_of_reviews last_review  \
23371  Entire home/apt   95.0             1.0              266.0  2019-06-28   
41945  Entire home/apt   95.0             2.0               27.0  2019-07-05   
42252     Private room  150.0             1.0               24.0  2019-06-20   

       reviews_per_month  calculated_host_listings_count  availability_365  
23371              10.34                             1.0             345.0  
41945               8.62                             1.0             153.0  
42252               8.37                             5.0             247.0

### Ejercicio 3:
- Quiero rentar un Airbnb completo por un total de 200 días, en Brooklyn o Manhattan; no quiero pagar más de 150 dólares por noche, ¿Cuáles son las 5 opciones más economicas que me ofrece Airbnb?

### Ejercicio 4:
- En un nuevo Dataframe muestra por municipio: el número de habitaciones tipo individual, casa o departamento entero y compartido; también muestra en otra columna el precio promedio de todas las habitaciones por municipio.

In [29]:
df.groupby(["municipio", "room_type"]).agg({"price": "mean", "id": "count"})

price     id
municipio     room_type                         
Bronx         Entire home/apt  127.506596    379
              Private room      66.788344    652
              Shared room       59.800000     60
Brooklyn      Entire home/apt  178.327545   9559
              Private room      76.500099  10132
              Shared room       50.527845    413
Manhattan     Entire home/apt  249.239109  13199
              Private room     116.776622   7982
              Shared room       88.977083    480
Queens        Entire home/apt  147.050573   2096
              Private room      71.762456   3372
              Shared room       69.020202    198
Staten Island Entire home/apt  173.846591    176
              Private room      62.292553    188
              Shared room       57.444444      9

In [ ]:
df

### Ejercicio 5:
- Crea una nueva columna en el DataFrame original, con el rating de acuerdo a la siguiente formula, donde las variables son: puntaje del municipio, número de reviews y disponibilidad de la habitación al año.

- Formula: Rating = (Puntaje de ciudad) + (número de review/ 629) + (disponibilidad / 365).

- El puntaje para cada ciudad es el siguiente {"Manhattan" : 10/10, "Brooklyn" : 8/10, "Queens" : 6/10, "Staten Island" : 4/10, "Bronx" : 2/10}.

In [31]:
def rating(row):
    puntajes = {"Manhattan" : 10/10, "Brooklyn" : 8/10, "Queens" : 6/10, "Staten Island" : 4/10, "Bronx" : 2/10}
    municipio = row['municipio']
    number_of_reviews = row['number_of_reviews']
    disponibilidad = row['availability_365']

    puntaje_municipio = puntajes[municipio]
    return puntaje_municipio + number_of_reviews / 629 + disponibilidad / 365


In [33]:
df['rating'] = df.apply(rating, axis = 1)

In [ ]:
df

### Ejercicio 6:
- Encuentra las 5 habitaciones de cada municipio con mayor rating.

In [39]:
df.sort_values("rating", ascending=False).groupby("municipio").head(5)

id                                               name   host_id  \
2030     903947                     Beautiful Bedroom in Manhattan   4734398   
2015     891117                       Private Bedroom in Manhattan   4734398   
2031     903972                         Great Bedroom in Manhattan   4734398   
11759   9145202                            Room near JFK Queen Bed  47621202   
195       51572                        Prime Location in Manhattan    237329   
225       58059                PRIVATE Room on Historic Sugar Hill    277379   
1547     699472                    NYC 1st Shipping Container Home   3587751   
20403  16276632  Cozy Room Family Home LGA Airport NO CLEANING FEE  26432133   
4870    3474320                 Private brownstone studio Brooklyn  12949460   
4931    3533604                 Private Room in a two bedroom apt.   9510645   
1103     467634                                   yahmanscrashpads   2321321   
3105    1843108  HappyCozy GuestSuite w/ Great Energy close to JFK   2021121   
1271     546383                   My Little Guest Room in Flushing   2680820   
1571     713538      Chic Victorian private apartment in townhouse   3672774   
8001    6157276                    Bright LARGE BED near Manhattan  23591164   
256       63320                 D Private Che@p Room 2 Explore NYC    303939   
249       62452                     A SpeciaL!! Private Room in NY    303939   
510      182177              A PRIVATE FLAT / APARTMENT- $SPECIAL$    873273   
23371  18914730                       IDEALLY LOCATED COZY COTTAGE   1776707   
251       62787                        C Private Room By The Ferry    303939   
13046   9882254                        Owl's Landing - Garden Room  50068597   
18699  14804661              Cozy 4 beds, Free Ferry to Manhattan.  92493393   
171       44096                                   Room with a View    190409   
1193     506575                   Bronx Room Near Yankees + Harlem   2001830   
14310  11138823                   A beautiful Cozy 1 BR Apartment.  57885474   

               host_name      municipio          barrio  latitude  longitude  \
2030                  Jj      Manhattan          Harlem  40.82124  -73.93838   
2015                  Jj      Manhattan          Harlem  40.82264  -73.94041   
2031                  Jj      Manhattan          Harlem  40.82085  -73.94025   
11759               Dona         Queens         Jamaica  40.66730  -73.76831   
195                  Lee      Manhattan         Chelsea  40.74859  -73.99671   
225                Agnes      Manhattan          Harlem  40.82510  -73.94287   
1547         Janet-David       Brooklyn    Williamsburg  40.70995  -73.95536   
20403           Danielle         Queens   East Elmhurst  40.76335  -73.87007   
4870                 Asa       Brooklyn      Park Slope  40.67926  -73.97711   
4931               Rylan       Brooklyn        Bushwick  40.70399  -73.92813   
1103               Lloyd         Queens         Jamaica  40.67747  -73.76493   
3105               Megan       Brooklyn   Cypress Hills  40.67840  -73.89362   
1271               Linda         Queens        Flushing  40.75578  -73.81948   
1571              Alison       Brooklyn    Clinton Hill  40.68341  -73.96060   
8001              Angela         Queens   East Elmhurst  40.76664  -73.87577   
256             Lissette  Staten Island   Tompkinsville  40.63481  -74.08519   
249             Lissette  Staten Island   Tompkinsville  40.63536  -74.08537   
510    Christian & Carla          Bronx        Allerton  40.86466  -73.85709   
23371               Luis          Bronx     Throgs Neck  40.83051  -73.82511   
251             Lissette  Staten Island   Tompkinsville  40.63518  -74.08546   
13046               Alex  Staten Island       Stapleton  40.63516  -74.07735   
18699       Eddie & Lois  Staten Island   West Brighton  40.63333  -74.11311   
171             Waundell          Bronx      Highbridge  40.83232  -73.93184   
1193     Jos

### Ejercicio 7:
- Crea un DataFrame con la cantidad de reviews de cada municipio de acuerdo a tipo de habitación.

In [ ]:
df.groupby(['municipio', 'room_type'], as_index=False)['number_of_reviews'].sum()

### Ejercicio 8:
- Encuentra los 10 barrios con mayor precio por habitación individual.

In [ ]:
df[df['room_type'] == "Private room"].groupby("barrio")['price'].mean().sort_values(ascending=False).head(10)

### Ejercicio 9:
- Encuentra el barrio con mejor rating de cada municipio.

In [55]:
def mejor_barrio(df):
    return df.groupby("barrio")['rating'].mean().sort_values(ascending=False).index[0]

In [58]:
mejor_barrio_series = df.groupby("municipio").apply(mejor_barrio)

In [59]:
mejor_barrio_series

municipio
Bronx                  Co-op City
Brooklyn               Mill Basin
Manhattan        Theater District
Queens                   Edgemere
Staten Island      Fort Wadsworth
dtype: object

### Ejercicio 10:
- Del resultado anterio, selecciona los 2 barrios con el menor precio por "dep/casa completo".

In [63]:
df[df['barrio'].isin(mejor_barrio_series.values)]

id                                               name    host_id  \
553      205043                            Modern Condo in Midtown    1007558   
567      212178                              1 Bedroom Pre War apt    1094178   
725      265145                    Studio sublet in Hell's kitchen    1390947   
2267    1080766                  Summer Special Price Times Square    5944682   
2717    1466552  ALL YOU NEED!! HK/TIME SQ./THEATER (READ REVIEWS)    7112116   
...         ...                                                ...        ...   
47913  35977480                                    The Empress den  270651073   
47956  36001552                             NYC Traveler’s dream  197551920   
48293  36182136                            House by the beach side  234781729   
48560  36315612  Convenient Midtown 1BR w/ Gym, W/D, Doorman, B...  107434423   
48616  36344347   Beautiful apartment /1 min walk from Time Square  119005168   

                 host_name  municipio            barrio  latitude  longitude  \
553    Welcome To My Place  Manhattan  Theater District  40.75895  -73.98830   
567                 Jeremy  Manhattan  Theater District  40.75877  -73.98863   
725                  Akiko  Manhattan  Theater District  40.76217  -73.98411   
2267         Maja & Pierre  Manhattan  Theater District  40.75964  -73.98534   
2717                    Ty  Manhattan  Theater District  40.76199  -73.98627   
...                    ...        ...               ...       ...        ...   
47913              Nazesta     Queens          Edgemere  40.59540  -73.77640   
47956                Mehek  Manhattan  Theater District  40.76122  -73.98655   
48293                Nikky     Queens          Edgemere  40.59309  -73.77958   
48560           Blueground  Manhattan  Theater District  40.76111  -73.98511   
48616                 Noah  Manhattan  Theater District  40.75982  -73.98662   

             room_type  price  minimum_nights  number_of_reviews last_review  \
553       Private room  150.0             1.0              330.0  2019-06-20   
567    Entire home/apt  230.0             2.0              179.0  2019-06-19   
725    Entire home/apt  200.0             5.0                8.0  2015-01-08   
2267      Private room  185.0             2.0              243.0  2019-06-27   
2717      Private room  169.0             2.0               33.0  2019-04-10   
...                ...    ...             ...                ...         ...   
47913     Private room   50.0             3.0                0.0         NaT   
47956     Private room   70.0             1.0                2.0  2019-07-07   
48293     Private room   65.0             1.0                2.0  2019-07-06   
48560  Entire home/apt  303.0            30.0                0.0         NaT   
48616  Entire home/apt  400.0             2.0                0.0         NaT   

       reviews_per_month  calculated_host_listings_count  availability_365  \
553                 7.14                             1.0             111.0   
567                 1.88                             1.0              44.0   
725                 0.09                             1.0               0.0   
2267                3.22                             1.0             127.0   
2717                0.47                             2.0             340.0   
...                  ...                             ...               ...   
47913                NaN                             1.0             363.0   
47956               2.00                             1.0              39.0   
48293               2.00                             1.0             363.0   
48560                NaN                           232.0             318.0   
48616                NaN                             1.0               3.0   

         rating  
553    1.828752  
567    1.405127  
725    1.012719  
2267   1.734273  
2717   1.983971  
...         ...  
47913  1.594521  
47956  1.110029  
48293  1.597700  
4